### PR TITLE
Color-change-v0.0

### PR DESCRIPTION
**Pull request by RathoreDevansh08**
Fixes convict-git/KoraKaagaz#35

### CODE REVIEW

**Comment by RathoreDevansh08** [comment]
To be reviewed by @wildcat26. Exception handling: to be done during test harness.

File: src/processing/ParameterizedOperationsUtil.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
+package processing;
+
+import java.util.*;
+import processing.utility.*;
+
+/**
+*
+* @author Devansh Singh Rathore
```
Comment (wildcat26) [inline src/processing/ParameterizedOperationsUtil.java:8]: Please add my name as the reviewer.

File: src/processing/ParameterizedOperationsUtil.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+*/
+public class ParameterizedOperationsUtil {
+
+    /**
+    * Function to update undo-redo stack. Undo-Redo module cannot call this function inorder to
+    * avoid logical errors.
+    */
+    private static void stackUtil(BoardObject newObj) {
+
+        UndoRedo.pushIntoStack(newObj);
```
Comment (wildcat26) [inline src/processing/ParameterizedOperationsUtil.java:24]: Need the appropriate import for UndoRedo. `UndoRedo` would be present in `src/processing/` or `src/processing/utility`

File: src/processing/ParameterizedOperationsUtil.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    */
+    public static BoardObject colorChangeUtil(BoardObject obj, UserId id, Intensity newIntensity) {
+
+        // storing initial intensities of pixels (to be used during undo-operation)
+        ArrayList<Pixel> prevPixelIntensities = new ArrayList<Pixel>();
+        prevPixelIntensities = obj.getPixels();
+
+        // changing intensities of selected object's pixels
+        ArrayList<Pixel> newPixelSet = new ArrayList<Pixel>();
+        newPixelSet = obj.getPixels();
```
Comment (wildcat26) [inline src/processing/ParameterizedOperationsUtil.java:42]: You already have `obj.getPixels()` in line 37 in the variable `prevPixelIntensities`. No need to call `getPixels` again.

File: src/processing/ParameterizedOperationsUtil.java (hunk 2 above)
Comment (wildcat26) [reply src/processing/ParameterizedOperationsUtil.java:24]: Please confirm.

File: src/processing/ParameterizedOperationsUtil.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+        // set (COLOR_CHANGE) as the operation which is applied on object.
+        BoardObjectOperation newBoardOp = new ColorChangeOperation(newIntensity);
+
+        // remove previous object from maps.
+        BoardObject dummyObj = SelectDeleteUtil.removeObjectFromMaps(obj.objectId);
+
+        // create a new object with same objectId, timestamp and other updated values.
+        BoardObject newObj = CurveBuilder.drawCurve(newPixelSet, newBoardOp, newObjectId, newTime,
+        id, prevPixelIntensities, false);
```
Comment (wildcat26) [inline src/processing/ParameterizedOperationsUtil.java:60]: why `prevPixelIntensities` here? Shouldn't it be `newPixelSet` ?

File: src/processing/ParameterizedOperationsUtil.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
```
Comment (wildcat26) [inline src/processing/ParameterizedOperationsUtil.java:61]: Just like you call `removeObjectFromMaps`, you also need to call `insertObjectIntoMaps (newObj)` (on the new board object) to update the maps

File: src/processing/ParameterizedOperationsUtil.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        return newObj;
+    }
+
+    /**
+    * Function for implementing color-change of select-able objects.
+    */
+    public static ObjectId colorChange(BoardObject obj, UserId id, Intensity intensity) {
```
Comment (wildcat26) [inline src/processing/ParameterizedOperationsUtil.java:69]: According to the interface provided to the UI, color change operation can only happen on the selected object. So we do not need to `BoardObject` argument here.

File: src/processing/ParameterizedOperationsUtil.java (hunk 3 above)
Comment (wildcat26) [reply src/processing/ParameterizedOperationsUtil.java:42]: `newPixelSet = prevPixelIntensities`

**Review by wildcat26 — CHANGES_REQUESTED** [review]

File: src/processing/ParameterizedOperationsUtil.java (hunk 4 above)
Comment (wildcat26) [reply src/processing/ParameterizedOperationsUtil.java:60]: Not sure about Satchit's APIs. where can I find them? In his spec, the API for `drawCurve` is totally different from how you have used it here.

File: src/processing/ParameterizedOperationsUtil.java (hunk 3 above)
Comment (RathoreDevansh08) [reply src/processing/ParameterizedOperationsUtil.java:42]: for easy in reading...complexity wise both are same, so...

File: src/processing/ParameterizedOperationsUtil.java (hunk 4 above)
Comment (RathoreDevansh08) [inline src/processing/ParameterizedOperationsUtil.java:61]: both `prevPixelIntensities` and `newPixelSet` are different arguments required for BoardObject creation. Satchit's API has to be changed. Confirmed on Discord(https://discordapp.com/channels/@me/747776066034204733/768366664404172860).

File: src/processing/ParameterizedOperationsUtil.java (hunk 3 above)
Comment (wildcat26) [reply src/processing/ParameterizedOperationsUtil.java:42]: I felt this one is easier in reading.. and looks cleaner. Your call finally.

File: src/processing/ParameterizedOperationsUtil.java (hunk 6 above)
Comment (RathoreDevansh08) [reply src/processing/ParameterizedOperationsUtil.java:69]: Himanshu recommended this change. He said that Client Side Processing will  ensure that only selected objects get passed to colorChange through input parameter.

File: src/processing/ParameterizedOperationsUtil.java (hunk 2 above)
Comment (RathoreDevansh08) [reply src/processing/ParameterizedOperationsUtil.java:24]: Rakesh confirmed.

File: src/processing/ParameterizedOperationsUtil.java (hunk 6 above)
Comment (wildcat26) [reply src/processing/ParameterizedOperationsUtil.java:69]: Okay. But when he calls this function to implement the interface, he should make sure that it gets called **only for the currently selected object**

**Review by wildcat26 — APPROVED** [review]

**Pull request merged**